lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -45,6 +50,38 @@ def __init__(self, config: PbenchServerConfig, logger: logging.Logger):
             role=API_OPERATION.READ,
         )
 
+    def get_paginated_obj(self, query: Query, json: JSON, url: str, page_limit: int):
+        """
+        Helper function to return a slice of dataset according to the page limit and
+        a page number and a json object containing next page url.
+        TODO: There are multiple ways to optimize the functionality
+            1. Use of caching mechanism to temporarily store the queried datasets
+            instead of querying the database with every page request.
+            2. Use of unique pointers to store the last returned row and then use
+            this pointer in subsequent page request instead of initial start to
+            narrow down the result.
```

Comment:
> I thought Redis would be able to handle that.

I didn't think that Redis was involved, here...did you mean Postgres or SQLAlchemy?

> Implementing this approach means sorting the dataset on a unique column preferably a created timestamp

We need to determine exactly what our requirements are.  For starters, as Dave pointed out, eventually we will presumably want the client to be able to specify the ordering.  Then we need to determine whether we're allowed to return "fuzzy" results (i.e., is it OK if we return fewer items than requested?...is it OK if we return more?).  Finally, we need to decide what sort of help the database query can provide, and how expensive those various options are.

For instance, if we allow the client to specify the sorting order but restrict it to a single column, then we can come up with a pair of values which bound the items returned (i.e., every item returned has a value in the key column which is greater than or equal to `MIN` and less than `MAX`), and we return the `MAX` value with the results so that it becomes the `MIN` for the next query.  The problem with this approach is that for a given range of `MIN..MAX`, the number of items may not be exactly equal to the client's requested page size...if there is more than one value, then it's straightforward to return fewer rows than what was requested, but if all the selected rows have the same value then we could easily end up needing to return way more than what was requested.  If this is OK, then we're golden...but, if we want something more conformant or deterministic, then we need to work harder (which is where your desire for unique key values comes in...).  (We can do better if the client specifies multiple columns, but then we have to track the `MAX` for each of the columns, and there's still no guarantee of uniqueness.)

If the database query provides a stable sort, then we can use `offset()` and `limit()` instead of first doing a min/max selection.  If this is reasonably efficient (e.g., if sorting and offsetting is comparable in cost to doing a selection first and sorting the result of that), then this approach is easy; however, if this gets painful as the number of datasets gets large, then we'll want to look at another approach.

So, this is a rather long-winded way of saying that what I think you're proposing here is an excellent initial approach.  We'll presumably want to tweak it to allow the client to specify the basis for the sort, but, until the database operation proves to be too expensive, I don't think we need to worry about alternatives.

Reply:
Do we need to retain this `TODO` comment in this change?

Reply:
Well, we would still need to update the algorithm (or optimize it) if we need to so that's there for a probable TODO. So I will just update the docstring a little bit.